File: rust/17.letter-combinations-of-a-phone-number.rs
/*
 * @lc app=leetcode id=17 lang=rust
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start
impl Solution {
    pub fn letter_combinations(digits: String) -> Vec<String> {
        let mut combinations: Vec<String> = vec![];
        if digits.len() > 0 {
            let map = [
                vec!['a', 'b', 'c'],
                vec!['d', 'e', 'f'],
                vec!['g', 'h', 'i'],
                vec!['j', 'k', 'l'],
                vec!['m', 'n', 'o'],
                vec!['p', 'q', 'r', 's'],
                vec!['t', 'u', 'v'],
                vec!['w', 'x', 'y', 'z'],
            ];
            fn backtrack_insert(
                next_digit: &str,
                map: &[Vec<char>; 8],
                combination: String,
                combinations: &mut Vec<String>,
            ) {
                if let Some(d) = next_digit.chars().nth(0) {
                    for c in map[d.to_digit(10).unwrap() as usize - 2].iter() {
                        backtrack_insert(
                            &next_digit[1..],
                            map,
                            format!("{combination}{c}"),
                            combinations,
                        );
                    }
                } else {
                    combinations.push(combination);
                }
            }
            backtrack_insert(&digits, &map, String::new(), &mut combinations);
        }
        return combinations;
    }
}
// @lc code=end

File: javascript/17.letter-combinations-of-a-phone-number.js
/*
 * @lc app=leetcode id=17 lang=javascript
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function (digits) {
  const combinations = [];
  if (digits.length > 0) {
    const map = {
      2: "abc",
      3: "def",
      4: "ghi",
      5: "jkl",
      6: "mno",
      7: "pqrs",
      8: "tuv",
      9: "wxyz",
    };
    function backtrackInsert(n, combination) {
      if (n === digits.length) combinations.push(combination);
      else {
        for (const c of map[digits[n]]) {
          backtrackInsert(n + 1, combination + c);
        }
      }
    }
    backtrackInsert(0, "");
  }
  return combinations;
};
// @lc code=end
